File: Python/2-TwoPointers/Problems/1-ValidPalindrome/ValidPalindrome.py
import re  # Import the regular expression module

# Valid Palindrome - Two Pointers Approaches
# The problem involves checking if a given string is a palindrome, considering only alphanumeric characters and ignoring cases.

# Approach 1: Two Pointers (Optimized for Space)
# Time Complexity: O(n), Space Complexity: O(1)
def is_palindrome_two_pointers(s):
    """
    Checks if a string is a palindrome using the two-pointers approach.

    Args:
        s (str): The input string.

    Returns:
        bool: True if the string is a palindrome, False otherwise.
    """
    left, right = 0, len(s) - 1  # Initialize two pointers, one at the start and one at the end
    while left < right:  # Continue as long as the left pointer is to the left of the right pointer
        while left < right and not s[left].isalnum():
            # Move the left pointer to the right until it points to an alphanumeric character
            left += 1
        while left < right and not s[right].isalnum():
            # Move the right pointer to the left until it points to an alphanumeric character
            right -= 1
        if s[left].lower() != s[right].lower():
            # If the characters at the two pointers (converted to lowercase) are not equal,
            # the string is not a palindrome
            return False
        left += 1  # Move the left pointer one step to the right
        right -= 1  # Move the right pointer one step to the left
    return True  # If the loop finishes without finding any mismatched characters, the string is a palindrome

# Approach 2: Using Regular Expressions and Reversed String
# Time Complexity: O(n), Space Complexity: O(n)
def is_palindrome_regex(s):
    """
    Checks if a string is a palindrome using regular expressions and string reversal.

    Args:
        s (str): The input string.

    Returns:
        bool: True if the string is a palindrome, False otherwise.
    """
    cleaned = re.sub(r'[^a-zA-Z0-9]', '', s).lower()  # Remove non-alphanumeric characters and convert to lowercase
    return cleaned == cleaned[::-1]  # Compare the cleaned string with its reversed version

# Approach 3: Using Stack (LIFO)
# Time Complexity: O(n), Space Complexity: O(n)
def is_palindrome_stack(s):
    """
    Checks if a string is a palindrome using a stack.

    Args:
        s (str): The input string.

    Returns:
        bool: True if the string is a palindrome, False otherwise.
    """
    stack = []  # Initialize an empty stack
    for char in s:
        if char.isalnum():
            stack.append(char.lower())  # Push alphanumeric characters (in lowercase) onto the stack
    # Now the stack contains the alphanumeric characters in forward order.
    for char in s:
        if char.isalnum() and (stack.pop() != char.lower()):
            return False
    return True

# Approach 4: Recursion (Not Recommended for Large Strings)
# Time Complexity: O(n), Space Complexity: O(n) (Due to recursive stack)
def is_palindrome_recursive(s, left=0, right=None):
    """
    Checks if a string is a palindrome using recursion.

    Args:
        s (str): The input string.
        left (int, optional): The left pointer. Defaults to 0.
        right (int, optional): The right pointer. Defaults to None.

    Returns:
        bool: True if the string is a palindrome, False otherwise.
    """
    if right is None:
        s = ''.join(c.lower() for c in s if c.isalnum())  # Clean the string and convert it to lowercase
        right = len(s) - 1  # Initialize the right pointer
    if left >= right:
        # Base case: if the left pointer crosses or meets the right pointer, it's a palindrome
        return True
    if s[left] != s[right]:
        # If the characters at the two pointers are not equal, it's not a palindrome
        return False
    # Recursive step: check the substring excluding the characters at the current pointers
    return is_palindrome_recursive(s, left + 1, right - 1)

# Approach 5: Deque (Efficient for Pop from Both Ends)
# Time Complexity: O(n), Space Complexity: O(n)
from collections import deque  # Import the deque class from the collections module

def is_palindrome_deque(s):
    """
    Checks if a string is a palindrome using a deque.

    Args:
        s (str): The input string.

    Returns:
        bool: True if the string is a palindrome, False otherwise.
    """
    d = deque()  # Initialize an empty deque
    for char in s:
        if char.isalnum():
            d.append(char.lower())  # Append alphanumeric characters (in lowercase) to the deque
    while len(d) > 1:
        # Continue as long as the deque has more than one element
        if d.popleft() != d.pop():
            # Compare the characters at both ends; if they are not equal, it's not a palindrome
            return False
    return True  # If the loop finishes without finding any mismatched characters, it's a palindrome

# Test Cases
test_cases = ["A man, a plan, a canal: Panama", "race a car", "", "abba", "No 'x' in Nixon"]

for i, test in enumerate(test_cases, 1):
    print(f"Test Case {i}: {test}")
    print("Two Pointers:", is_palindrome_two_pointers(test))
    print("Regex Method:", is_palindrome_regex(test))
    print("Stack Method:", is_palindrome_stack(test))
    print("Recursive Method:", is_palindrome_recursive(test))
    print("Deque Method:", is_palindrome_deque(test))
    print()

File: Java/2-TwoPointers/Problems/1-ValidPalindrome/ValidPalindrome.java
import java.util.Locale;

public class ValidPalindrome {

    public static void main(String[] args) {
        String testString1 = "A man, a plan, a canal: Panama";
        String testString2 = "race a car";
        String testString3 = " ";
        String testString4 = "Madam, I'm Adam!";
        String testString5 = "121";

        // Test cases for each approach
        System.out.println("Test Cases for Valid Palindrome:");
        System.out.println("String 1: " + testString1);
        System.out.println("String 2: " + testString2);
        System.out.println("String 3: " + testString3);
        System.out.println("String 4: " + testString4);
        System.out.println("String 5: " + testString5);
        System.out.println("\nResults:");

        // Approach 1: Using StringBuilder (Two Pointers)
        System.out.println("\nApproach 1: Using StringBuilder (Two Pointers)");
        System.out.println("String 1: " + isPalindrome1(testString1));
        System.out.println("String 2: " + isPalindrome1(testString2));
        System.out.println("String 3: " + isPalindrome1(testString3));
        System.out.println("String 4: " + isPalindrome1(testString4));
        System.out.println("String 5: " + isPalindrome1(testString5));


        // Approach 2: Using Two Pointers (Optimized)
        System.out.println("\nApproach 2: Using Two Pointers (Optimized)");
        System.out.println("String 1: " + isPalindrome2(testString1));
        System.out.println("String 2: " + isPalindrome2(testString2));
        System.out.println("String 3: " + isPalindrome2(testString3));
        System.out.println("String 4: " + isPalindrome2(testString4));
        System.out.println("String 5: " + isPalindrome2(testString5));

        // Approach 3: Using Recursion
        System.out.println("\nApproach 3: Using Recursion");
        System.out.println("String 1: " + isPalindrome3(testString1));
        System.out.println("String 2: " + isPalindrome3(testString2));
        System.out.println("String 3: " + isPalindrome3(testString3));
        System.out.println("String 4: " + isPalindrome3(testString4));
        System.out.println("String 5: " + isPalindrome3(testString5));

        // Approach 4: Using Regular Expression and StringBuilder
        System.out.println("\nApproach 4: Using Regular Expression and StringBuilder");
        System.out.println("String 1: " + isPalindrome4(testString1));
        System.out.println("String 2: " + isPalindrome4(testString2));
        System.out.println("String 3: " + isPalindrome4(testString3));
        System.out.println("String 4: " + isPalindrome4(testString4));
        System.out.println("String 5: " + isPalindrome4(testString5));

        // Approach 5: Using Character.isLetterOrDigit and Two Pointers
        System.out.println("\nApproach 5: Using Character.isLetterOrDigit and Two Pointers");
        System.out.println("String 1: " + isPalindrome5(testString1));
        System.out.println("String 2: " + isPalindrome5(testString2));
        System.out.println("String 3: " + isPalindrome5(testString3));
        System.out.println("String 4: " + isPalindrome5(testString4));
        System.out.println("String 5: " + isPalindrome5(testString5));
    }

    /**
     * Approach 1: Using StringBuilder (Two Pointers)
     *
     * - Filters out non-alphanumeric characters and converts the string to lowercase.
     * - Uses StringBuilder to create a reversed version of the filtered string.
     * - Compares the filtered string with its reversed version.
     *
     * Time Complexity: O(n), where n is the length of the string.
     * Space Complexity: O(n), for the StringBuilder.
     */
    public static boolean isPalindrome1(String s) {
        if (s == null) {
            return true; // Handle null input as a special case.
        }
        StringBuilder filteredString = new StringBuilder();
        for (char c : s.toCharArray()) {
            if (Character.isLetterOrDigit(c)) {
                filteredString.append(Character.toLowerCase(c));
            }
        }
        String reversedString = filteredString.reverse().toString();
        return filteredString.toString().equals(reversedString);
    }

    /**
     * Approach 2: Using Two Pointers (Optimized)
     *
     * - Uses two pointers, one starting from the beginning and the other from the end.
     * - Iterates through the string, comparing characters at the two pointer positions.
     * - Skips non-alphanumeric characters.
     * - More efficient as it avoids creating new strings.
     *
     * Time Complexity: O(n), where n is the length of the string.
     * Space Complexity: O(1), constant space.
     */
    public static boolean isPalindrome2(String s) {
        if (s == null) {
            return true;
        }
        int left = 0;
        int right = s.length() - 1;
        while (left < right) {
            while (left < right && !Character.isLetterOrDigit(s.charAt(left))) {
                left++;
            }
            while (left < right && !Character.isLetterOrDigit(s.charAt(right))) {
                right--;
            }
            if (left < right && Character.toLowerCase(s.charAt(left)) != Character.toLowerCase(s.charAt(right))) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }

    /**
     * Approach 3: Using Recursion
     *
     * - Recursively checks if the substring is a palindrome.
     * - Skips non-alphanumeric characters using helper methods.
     * - Base case: If the substring has 0 or 1 character, it's a palindrome.
     *
     * Time Complexity: O(n), where n is the length of the string.
     * Space Complexity: O(n), due to the recursion stack.
     */
    public static boolean isPalindrome3(String s) {
        if (s == null) {
            return true;
        }
        String filteredString = "";
        for(char c: s.toCharArray()){
            if(Character.isLetterOrDigit(c)){
                filteredString += c;
            }
        }
        filteredString = filteredString.toLowerCase();
        return isPalindromeRecursive(filteredString, 0, filteredString.length() - 1);
    }

    private static boolean isPalindromeRecursive(String s, int left, int right) {
        if (left >= right) {
            return true;
        }
        if (s.charAt(left) != s.charAt(right)) {
            return false;
        }
        return isPalindromeRecursive(s, left + 1, right - 1);
    }



    /**
     * Approach 4: Using Regular Expression and StringBuilder
     *
     * - Uses a regular expression to remove non-alphanumeric characters.
     * - Converts the string to lowercase.
     * - Uses StringBuilder to reverse the cleaned string and compare.
     *
     * Time Complexity: O(n), where n is the length of the string.
     * Space Complexity: O(n), for the StringBuilder.
     */
    public static boolean isPalindrome4(String s) {
        if (s == null) {
            return true;
        }
        String cleanedString = s.toLowerCase().replaceAll("[^a-z0-9]", "");
        String reversedString = new StringBuilder(cleanedString).reverse().toString();
        return cleanedString.equals(reversedString);
    }

    /**
     * Approach 5: Using Character.isLetterOrDigit and Two Pointers
     *
     * - This is a refined version of the two-pointer approach.
     * - It directly uses Character.isLetterOrDigit to check characters.
     * - It converts characters to lowercase for comparison.
     * - Efficient and avoids unnecessary string creation.
     *
     * Time Complexity: O(n), where n is the length of the string.
     * Space Complexity: O(1), constant space.
     */
    public static boolean isPalindrome5(String s) {
        if (s == null) {
            return true;
        }
        int left = 0;
        int right = s.length() - 1;

        while (left < right) {
            // Move left pointer to the next alphanumeric character
            while (left < right && !Character.isLetterOrDigit(s.charAt(left))) {
                left++;
            }
            // Move right pointer to the previous alphanumeric character
            while (left < right && !Character.isLetterOrDigit(s.charAt(right))) {
                right--;
            }
            // Compare characters at both pointers (case-insensitive)
            if (left < right && Character.toLowerCase(s.charAt(left)) != Character.toLowerCase(s.charAt(right))) {
                return false; // If characters don't match, it's not a palindrome
            }
            // Move pointers towards the center
            left++;
            right--;
        }
        return true; // If the loop finishes without finding a mismatch, it's a palindrome
    }
}

